### PR TITLE
Update paraguay scrapers

### PR DESCRIPTION
**Pull request by aguilerapy**
#323
- [x] fix middlewares
- [x] move logger
- [x] fix clientSecret value

### CODE REVIEW

File: kingfisher_scrapy/spiders/paraguay_hacienda.py [hunk 1]
```diff
@@ -13,17 +13,18 @@ class ParaguayHacienda(BaseSpider):
...
     release_ids = []
-    request_limit = 10000
     request_time_limit = 14.0
 
     custom_settings = {
         'DOWNLOADER_MIDDLEWARES': {
            'kingfisher_scrapy.middlewares.ParaguayAuthMiddleware': 543,
         },
         'CONCURRENT_REQUESTS': 1,
+        'DUPEFILTER_DEBUG': True,
```
Comment (jpmckinney) [inline kingfisher_scrapy/spiders/paraguay_hacienda.py]: Do we want this, or was it only here for testing?

File: kingfisher_scrapy/middlewares.py [hunk 2]
```diff
@@ -66,16 +66,21 @@ def process_request(self, request, spider):
             spider.crawler.stop()
             raise scrapy.exceptions.IgnoreRequest()
         request.headers['Authorization'] = spider.access_token
-        if self._expires_soon(spider):
+        if self._expires_soon(spider) or not spider.access_token:
```
Comment (jpmckinney) [inline kingfisher_scrapy/middlewares.py]: Why not put the `not spider.access_token` logic in `_expires_soon`?

File: kingfisher_scrapy/middlewares.py [hunk 3]
```diff
@@ -66,16 +66,21 @@ def process_request(self, request, spider):
...
+            # SAVE the last request to continue after getting the token
+            spider.last_request = request
```
Comment (jpmckinney) [inline kingfisher_scrapy/middlewares.py]: Isn't it possible for multiple simultaneous requests to meet the expiry condition? If so, should this be an array or requests to re-perform?

File: kingfisher_scrapy/spiders/paraguay_base.py [hunk 4]
```diff
@@ -48,10 +49,9 @@ def from_crawler(cls, crawler, *args, **kwargs):
         return spider
 
     def start_requests(self):
-        # Start request access token
-        self.request_access_token()
         yield scrapy.Request(
             self.base_page_url,
+            dont_filter=True,
```
Comment (jpmckinney) [inline kingfisher_scrapy/spiders/paraguay_base.py]: Can you add a comment explaining in what case we might send duplicate requests?

**Comment by jpmckinney** [comment]
It's not part of the PR, but I was very confused by what this line was trying to do:

```python
if 'auth' in request.meta and request.meta['auth'] is not None and not request.meta['auth']:
```

It would be clearer for this to be:

```python
if 'auth' in request.meta and request.meta['auth'] is False:
```

File: kingfisher_scrapy/spiders/paraguay_hacienda.py (hunk 1 above)
Comment (aguilerapy) [reply kingfisher_scrapy/spiders/paraguay_hacienda.py]: I think we can remove it from both scrapers.

File: kingfisher_scrapy/middlewares.py (hunk 3 above)
Comment (aguilerapy) [reply kingfisher_scrapy/middlewares.py]: At this time, we defined `'CONCURRENT_REQUESTS': 1` in `custom_settings`, and that multiple expiry case couldn't happen. If we want to have simultaneous requests, the code must be analyzed for that.

File: kingfisher_scrapy/middlewares.py [hunk 5]
```diff
@@ -88,4 +88,5 @@ def process_response(self, request, response, spider):
     @staticmethod
     def _expires_soon(spider):
         # spider MUST implement the expires_soon method
-        return spider.expires_soon(datetime.now() - spider.start_time) if spider.start_time else True
+        return spider.expires_soon(datetime.now() - spider.start_time) \
+            if (spider.start_time and spider.access_token) else True
```
Comment (jpmckinney) [inline kingfisher_scrapy/middlewares.py]: Please avoid one-line conditional expressions. Multiline if/else statements are much easier to read.

**Comment by aguilerapy** [comment]
Actually, with `dont_filter`, scrappy doesn't filter the request and doesn't be logged, `dont_filter` works in requests for access tokens and in the continuation of last_request saved. We have INFO: "Time_diff", "Requesting access token" and "New access token" for logs in those cases. @jpmckinney

**Comment by yolile** [comment]
@jpmckinney according to  [this](https://github.com/open-contracting/kingfisher-scrape/issues/323#issuecomment-594165048) comment we decided to keep using `crawler.crawl` but is ok to use this other solution that @aguilerapy found then?

**Comment by jpmckinney** [comment]
I think either option is okay.

The option in this PR won't work if we removed `'CONCURRENT_REQUESTS': 1`. Do you know if `crawler.crawl` can work with concurrent requests? If not, they we can leave it as-is.

**Comment by yolile** [comment]
> I think either option is okay.
> 
> The option in this PR won't work if we removed `'CONCURRENT_REQUESTS': 1`. Do you know if `crawler.crawl` can work with concurrent requests? If not, then we can leave it as-is.

Both options works only with CONCURRENT_REQUESTS set to 1

**Comment by yolile** [comment]
@jpmckinney can we merge this?

**Comment by jpmckinney** [comment]
Yes, it's approved :) I'm trying to fix Travis this week, so I'll merge in the meantime.

**Pull request merged**